internal/controller/nicconfigurationtemplate_controller.go

```diff
@@ -43,26 +57,254 @@ type NicConfigurationTemplateReconciler struct {
+					log.Log.Error(err, "Failed to drop device name from template status", "template", template)
+				}
+
+				continue
+			}
+
+			matchingTemplates = append(matchingTemplates, &template)
+		}
+
+		if len(matchingTemplates) == 0 {
+			log.Log.V(2).Info("Device doesn't match any configuration template, resetting the spec", "device", device.Name)
+			device.Spec.Configuration = nil
+			err = r.Update(ctx, &device)
+			if err != nil {
+				log.Log.Error(err, "Failed to update device's spec", "device", device)
+				return ctrl.Result{}, err
+			}
+			continue
+		} else if len(matchingTemplates) > 1 {
+			for _, template := range matchingTemplates {
+				err := r.dropDeviceFromStatus(ctx, device.Name, template)
+				if err != nil {
+					log.Log.Error(err, "Failed to drop device name from template status", "template", template)
+				}
+			}
 
-	// TODO(user): your logic here
+			templateNames := []string{}
+			for _, template := range matchingTemplates {
+				templateNames = append(templateNames, template.Name)
+			}
+			joinedTemplateNames := strings.Join(templateNames, ",")
+			err = fmt.Errorf("device matches several configuration templates: %s, %s", device.Name, joinedTemplateNames)
+			log.Log.Error(err, "Device matches several configuration templates, deleting its spec and emitting an error event", "device", device.Name)
+			return ctrl.Result{}, r.handleErrorSeveralMatchingTemplates(ctx, &device, joinedTemplateNames)
+		}
+
+		matchingTemplate := matchingTemplates[0]
+
+		if !slices.Contains(matchingTemplate.Status.NicDevices, device.Name) {
+			matchingTemplate.Status.NicDevices = append(matchingTemplate.Status.NicDevices, device.Name)
+			err = r.Status().Update(ctx, matchingTemplate)
+			if err != nil {
+				log.Log.Error(err, "Failed to update template's status", "template", matchingTemplate)
+				return ctrl.Result{}, err
+			}
+		}
+
+		err = r.applyTemplateToDevice(ctx, &device, matchingTemplates[0])
+		if err != nil {
+			return ctrl.Result{}, nil
+		}
+
+	}
 
 	return ctrl.Result{}, nil
 }
 
+func (r *NicConfigurationTemplateReconciler) dropDeviceFromStatus(ctx context.Context, deviceName string, template *v1alpha1.NicConfigurationTemplate) error {
+	if slices.Contains(template.Status.NicDevices, deviceName) {
+		// Device no longer matches template, drop it from the template's status
+		index := slices.Index(template.Status.NicDevices, deviceName)
+		template.Status.NicDevices = slices.Delete(template.Status.NicDevices, index, index+1)
+		return r.Status().Update(ctx, template)
+	}
+	return nil
+}
+
+func (r *NicConfigurationTemplateReconciler) applyTemplateToDevice(ctx context.Context, device *v1alpha1.NicDevice, template *v1alpha1.NicConfigurationTemplate) error {
+	log.Log.V(2).Info(fmt.Sprintf("Applying template %s to device %s", template.Name, device.Name))
+
+	updateSpec := false
+	if device.Spec.Configuration == nil {
+		updateSpec = true
+		device.Spec.Configuration = &v1alpha1.NicDeviceConfigurationSpec{}
```

Comment:
do we need to continue to the other if conditions ?

Reply:
Yes we do because NicDevice and NicConfigurationTemplate store different types in the root spec object and we have to compare / apply template field by field